collaborative-plonk/src/round3.rs

```diff
@@ -0,0 +1,602 @@
+        let two = P::ScalarField::one() + P::ScalarField::one();
+        let neg_eight = -(two.square() * two);
+        let root_of_unity = domains.roots_of_unity[2];
+        let two_root_unity = two * root_of_unity;
+        [zero, two + two_root_unity, neg_eight, two - two_root_unity]
+    }
+    fn compute_t(
+        driver: &mut T,
+        domains: &Domains<P>,
+        challenges: &Round3Challenges<T, P>,
+        zkey: &ZKey<P>,
+        polys: &Round2Polys<T, P>,
+    ) -> PlonkProofResult<[FieldShareVec<T, P>; 3]> {
+        let z1 = Self::get_z1(domains);
+        let z2 = Self::get_z2(domains);
+        let z3 = Self::get_z3(domains);
+        let mut w = P::ScalarField::one();
+        let mut ap = Vec::with_capacity(zkey.domain_size * 4);
+        let mut bp = Vec::with_capacity(zkey.domain_size * 4);
+        let mut cp = Vec::with_capacity(zkey.domain_size * 4);
+
+        let pow_root_of_unity = domains.roots_of_unity[zkey.power];
+        let pow_plus2_root_of_unity = domains.roots_of_unity[zkey.power + 2];
+        for _ in 0..zkey.domain_size * 4 {
+            //TODO use add_mul convenience method
+            let ap_ = driver.mul_with_public(&w, &challenges.b[0]);
+            let ap_ = driver.add(&challenges.b[1], &ap_);
+            ap.push(ap_);
+
+            let bp_ = driver.mul_with_public(&w, &challenges.b[2]);
+            let bp_ = driver.add(&challenges.b[3], &bp_);
+            bp.push(bp_);
+
+            let cp_ = driver.mul_with_public(&w, &challenges.b[4]);
+            let cp_ = driver.add(&challenges.b[5], &cp_);
+            cp.push(cp_);
+
+            w *= &pow_plus2_root_of_unity;
+        }
+
+        let ap_vec = ap.into();
+        let bp_vec = bp.into();
+        let cp_vec = cp.into();
+
+        // TODO parallelize these?
+        let a_b = driver.mul_vec(&polys.poly_eval_a.eval, &polys.poly_eval_b.eval)?;
+        let a_bp = driver.mul_vec(&polys.poly_eval_a.eval, &bp_vec)?;
+        let ap_b = driver.mul_vec(&polys.poly_eval_b.eval, &ap_vec)?;
+        let ap_bp = driver.mul_vec(&ap_vec, &bp_vec)?;
+
+        let mut e1 = Vec::with_capacity(zkey.domain_size * 4);
+        let mut e1z = Vec::with_capacity(zkey.domain_size * 4);
+
+        let mut e2a = Vec::with_capacity(zkey.domain_size * 4);
+        let mut e2b = Vec::with_capacity(zkey.domain_size * 4);
+        let mut e2c = Vec::with_capacity(zkey.domain_size * 4);
+        let mut e2d = Vec::with_capacity(zkey.domain_size * 4);
+        let mut zp = Vec::with_capacity(zkey.domain_size * 4);
+
+        let mut e3a = Vec::with_capacity(zkey.domain_size * 4);
+        let mut e3b = Vec::with_capacity(zkey.domain_size * 4);
+        let mut e3c = Vec::with_capacity(zkey.domain_size * 4);
+        let mut e3d = Vec::with_capacity(zkey.domain_size * 4);
+        let mut zwp = Vec::with_capacity(zkey.domain_size * 4);
+        let mut w = P::ScalarField::one();
+        for i in 0..zkey.domain_size * 4 {
```

Comment:
It is important for this loop to not have any MPC operations that need to communicate. IDK how to best enforce this in the current design, but we should at least comment on it.

Reply:
commented and enforced with for_each loop